### ISSUE TITLE
Convert documentation examples into testsuite tests

### ISSUE THREAD

**Issue by lgritz** [AcademySoftwareFoundation/OpenImageIO#3992]
The [documentation](https://openimageio.readthedocs.io/) contains lots of code examples embedded in the text. Some of them have bugs or are incorrect in various ways, either because they were never correct (having been typed in for the docs themselves and never directly tested), or were once correct but now the APIs have changed and the example was never updated to match.

To combat this, we would like to go through all the documentation, and to the greatest extent possible, convert every significant code example in the docs into a test -- that is, into actual code that will compile and run every time we run the testsuite -- and then the documentation will merely include the test code by reference, so they can never get out of sync again.

Explanation: https://github.com/OpenImageIO/oiio/wiki/Converting-documentation-examples-to-tests

There are dozens and dozens of code examples in the docs that we could do this with. Small steps -- do this for one or two at a time, that's fine.

**Comment by aryangavit** [comment]
Hey I would like to work on this, can you assign me this issue.

**Comment by lgritz** [comment]
I'm not going to formally assign the issue to anyone, because no one's piece of it is going to close the issue (until there are literally no more examples left to convert in the docs). This will take probably dozens of PRs (ad people, maybe) before it's truly complete.

Please feel free to work on it. Maybe leave a comment here saying which particular example(s) you want to do, so that others who also want to contribute to the task won't pick the same specific bit that you did.

**Comment by aryangavit** [comment]
> I'm not going to formally assign the issue to anyone, because no one's piece of it is going to close the issue (until there are literally no more examples left to convert in the docs). This will take probably dozens of PRs (ad people, maybe) before it's truly complete.
> 
> Please feel free to work on it. Maybe leave a comment here saying which particular example(s) you want to do, so that others who also want to contribute to the task won't pick the same specific bit that you did.

Hi, I'm a beginner in open source and excited to learn about testsuits, I'll learn more about them and hoping to give it my all in fixing the documentation. If you guys don't mind I'll start with `Image Output: Writing Images` section code snippets.

**Comment by lgritz** [comment]
I would recommend that anybody wanting to leave a note here declaring that they want to work on an example pick ONE specific code example to start with, and drive that to a complete accepted PR, before declaring a whole section or chapter off-limits to others. Get the hang of the process with something small, figure out how rapidly you can do the conversions, and then for round 2, you can reserve a larger section that might be a few days of work for you.

**Comment by buddly27** [comment]
Hi @lgritz, I'm interested in picking up the conversion for the examples within [imagebufalgo.rst](https://github.com/AcademySoftwareFoundation/OpenImageIO/blob/master/src/doc/imagebufalgo.rst) as part of [Dev Days](https://github.com/AcademySoftwareFoundation/OpenImageIO/wiki/Dev-Days-2023). As a OIIO newbie, it seems like an excellent opportunity to get to know the API better!

**Comment by lgritz** [comment]
That's great. It's a big chapter with a lot of example, maybe you want to claim a particular section of that chapter to start with?

**Comment by buddly27** [comment]
I can start with [common principles](https://github.com/buddly27/OpenImageIO/blob/master/src/doc/imagebufalgo.rst#imagebufalgo-common-principles) and move to [pattern generation](https://github.com/buddly27/OpenImageIO/blob/master/src/doc/imagebufalgo.rst#pattern-generation) if it goes smoothly. Unless someone else wants to get it :)

**Comment by lgritz** [comment]
You claimed it first, it's yours!

I think that for examples where in the docs we show images, what we'll want to do is also have the docs switch to actually using the reference images from the examples. If that seems like too much to do all at once, you don't have to do it all in one step. We can move the code first, then change the location of the images.

**Comment by MichaelRCarroll-Intel** [comment]
Hi @buddly27 and @lgritz... I'm on for dev days as well and new to the source. I'll take [ImageBuf: Image Buffers](https://openimageio.readthedocs.io/en/master/imagebuf.html#) to start. It has 3 (and a half) examples. After that, I could grab a section of imagebufalgo or perhaps another doc area that has an example1() template stub...

@lgritz  On the half example... The last "example" has an elipsis section by design ("same as above") ... It is used to illustrate an option for code abstraction... Should I assume this should best be positioned in the repository with the full example?

@lgritz On reference images, do you mean the hrefs in the documentation rst markdown that point to generated images that live in this [folder](https://github.com/AcademySoftwareFoundation/OpenImageIO/tree/master/testsuite/docs-examples-cpp/ref) or this [folder](https://github.com/AcademySoftwareFoundation/OpenImageIO/tree/master/testsuite/docs-examples-python/ref)

Thanks

**Comment by lgritz** [comment]
Sounds great. Start with that chapter, see how it goes. (Sometimes fooling around with the docs and the testsuite can take longer than expected.)

The ellipsis is just to make the code example not be too long by mostly repeating what is directly above it. I suppose that in that final example, it didn't need to show make_black_impl at all. Looking at the examples I did in the other chapter, you should be able to see how to use the `:start-after:` and `:end-before:` to clip particular regions to appear in the docs. So the testsuite code can be a full working example, with only the key section actually appearing in the docs.

Yeah, by reference images, I mean in the ref/ directory of each test are the sample output that is the standard for the test (i.e., "if we run the test and it matches the reference output, the test passes; if it differs, the test fails"). If the docs are clarified by showing the image, than we may also reference the image in the docs themselves, but not all code examples are clarified by also showing specific image output.

**Comment by CheeksTheGeek** [comment]
Hi @lgritz @MichaelRCarroll-Intel @buddly27 and everyone else planning to take a slice of this mega task or just wanting to browse the examples! I prepared a GitHub Project to track who's taking which example [here](https://github.com/orgs/AcademySoftwareFoundation/projects/22) and the overall progress. Since people will have different times and different sizes for this contribution, I made one for each example I could find. Will try to keep it updated with what you guys are doing. 

If possible, when someone is picking up a task or made progress on one or there's an example not in this list and you want to add it, ping me to add it to this project, and I'll update it asap.

**Comment by CheeksTheGeek** [comment]
For Dev Days, I'll start working [ImageOutput](https://openimageio.readthedocs.io/en/master/imageoutput.html) since the first two, [Simple Write](https://github.com/orgs/AcademySoftwareFoundation/projects/22/views/1?pane=issue&itemId=41187847) and [Writing Scanline](https://github.com/orgs/AcademySoftwareFoundation/projects/22/views/1?pane=issue&itemId=41187876) were covered in the [Prototype PR](https://github.com/AcademySoftwareFoundation/OpenImageIO/pull/3977), I'll continue from [Write Individual Tiles](https://github.com/orgs/AcademySoftwareFoundation/projects/22/views/1?pane=issue&itemId=41313856)

**Comment by buddly27** [comment]
Good morning everyone! I started a draft PR with one converted example to ensure we are all on the same page. I will update it and rebase my commit with the remaining examples from the "imagebufalgo" section throughout the day. Please let me know if you spot any issues!

@lgritz Would you prefer the PR to target an intermediate branch to ease collaborative work, or is it fine to target master?

**Comment by CheeksTheGeek** [comment]
Hi, @buddly27 for your `testsuite/docs-examples-cpp/out.txt` is there an OK just before it?, that's causing a test failure for the whole of **docs-examples-cpp** test when I was trying to test my local test additions.

```txt
OK
Comparing "simple.tif" and "ref/simple.tif"
PASS
Comparing "scanlines.tif" and "ref/scanlines.tif"
PASS
```

**Comment by buddly27** [comment]
Hi @CheeksTheGeek, I don't see the error, but maybe I'm doing something wrong. I'm running the tests using `ctest` to be more granular:
```console
OpenImageIO_ROOT=$(pwd)/dist LD_LIBRARY_PATH=$(pwd)/dist/lib/:$(pwd)/ext/dist/lib/ \
PYTHONPATH=$(pwd)/build/lib/python/site-packages/ \
ctest --test-dir ./build/ -R docs-examples- -VV
```

**Comment by CheeksTheGeek** [comment]
I do it using 
```sh
  ~/projects/oiio-build ❯ ctest docs-examples-cpp --rerun-failed --output-on-failure    5s  base  14:46:24
Test project /Users/$USER/projects/oiio-build
    Start 2: docs-examples-cpp
1/1 Test #2: docs-examples-cpp ................***Failed    2.91 sec
Comparing "simple.tif" and "ref/simple.tif"
PASS
Comparing "scanlines.tif" and "ref/scanlines.tif"
PASS
-rw-r--r--@ 1 $USER  staff  107 Oct 12 15:54 out.txt
-rw-r--r--@ 1 $USER  staff  104 Sep 27 19:02 ref/out.txt
command = cmake /Users/$USER/projects/oiio/testsuite/docs-examples-cpp -DCMAKE_BUILD_TYPE=Release >> build.txt 2>&1 ;
cmake --build . --config Release >> build.txt 2>&1 ;
./docs-examples-imageioapi >> out.txt  ;
./docs-examples-imageoutput >> out.txt  ;
./docs-examples-imageinput >> out.txt  ;
./docs-examples-writingplugins >> out.txt  ;
./docs-examples-imagecache >> out.txt  ;
./docs-examples-texturesys >> out.txt  ;
./docs-examples-imagebuf >> out.txt  ;
./docs-examples-imagebufalgo >> out.txt  ;

#### Error: this command failed:  cmake --build . --config Release >> build.txt 2>&1
FAIL
comparisons are ['ref/simple.tif']
comparing simple.tif to ref/simple.tif
PASS: simple.tif matches ref/simple.tif
comparisons are ['ref/scanlines.tif']
comparing scanlines.tif to ref/scanlines.tif
PASS: scanlines.tif matches ref/scanlines.tif
comparisons are ['ref/out.txt']
comparing out.txt to ref/out.txt
Diff out.txt vs ref/out.txt was:
-------
--- out.txt     Thu Oct 12 15:54:44 2023
+++ ref/out.txt Wed Sep 27 19:02:06 2023
@@ -1,4 +1,3 @@
-OK
 Comparing "simple.tif" and "ref/simple.tif"
 PASS
 Comparing "scanlines.tif" and "ref/scanlines.tif"

NO MATCH for out.txt
FAIL out.txt
-----out.txt----->
OK
Comparing "simple.tif" and "ref/simple.tif"
PASS
Comparing "scanlines.tif" and "ref/scanlines.tif"
PASS
<----------
-----ref/out.txt----->
Comparing "simple.tif" and "ref/simple.tif"
PASS
Comparing "scanlines.tif" and "ref/scanlines.tif"
PASS
<----------
Diff was:
-------
--- out.txt     Thu Oct 12 15:54:44 2023
+++ ref/out.txt Wed Sep 27 19:02:06 2023
@@ -1,4 +1,3 @@
-OK
 Comparing "simple.tif" and "ref/simple.tif"
 PASS
 Comparing "scanlines.tif" and "ref/scanlines.tif"



0% tests passed, 1 tests failed out of 1

Total Test time (real) =   2.94 sec

The following tests FAILED:
          2 - docs-examples-cpp (Failed)
Errors while running CTest
```

**Comment by lgritz** [comment]
> @lgritz Would you prefer the PR to target an intermediate branch to ease collaborative work, or is it fine to target master?

Target master is fine. I'll backport to the 2.5 branch later.

**Comment by buddly27** [comment]
@CheeksTheGeek Sorry for the late reply, as discussed in the [slack channel](https://academysoftwarefdn.slack.com/archives/C05782U3806/p1697082795107639), it might be some linking issue while attempting to run ctest outside of the wrapper (`make test TEST=doc` should hopefully work fine).

@lgritz I think #4016 is ready for review. I can tackle the next section separately to prevent overloading the review process!

**Comment by CalvinSch** [comment]
Hello everyone! I'm hopping in on this issue and hoping to tackle some [ImageInput](https://openimageio.readthedocs.io/en/v2.5.4.0/imageinput.html#image-input-made-simple) test cases, starting with [SimpleRead](https://github.com/orgs/AcademySoftwareFoundation/projects/22/views/1?pane=issue&itemId=41186535).

I'm new to the source and will be sure to reach out on Slack with more questions but wanted to get in on this discussion to help with planning.

**Comment by lgritz** [comment]
Just bumping this issue and reminding people that this is an ideal Dev Days task.

**Comment by benkant** [comment]
Hi @lgritz/ I'm in and am attempting to recruit some colleagues. I've registered and will claim some examples once I've had a closer look. Thanks for the excellent examples.

**Comment by lydia-zheng** [comment]
> Just bumping this issue and reminding people that this is an ideal Dev Days task.

Hi Larry! Registered and would be interested in working on this issue for Dev Day :)

**Comment by ziadkhouri** [comment]
Hi @lgritz , I've had a go at `imageinput-scanlines` https://github.com/AcademySoftwareFoundation/OpenImageIO/pull/4444

Hope that OK. z

**Comment by grdanny** [comment]
let me do the Image Enhancement / Restoration
https://openimageio.readthedocs.io/en/v2.5.15.0/imagebufalgo.html#image-enhancement-restoration

the fixNonFinite specifically is something that comes up a lot in my day to day life.

**Comment by lgritz** [comment]
Poking at this issue to make sure it's been updated recently to be included in clotributor.

This continues to be a great source of Dev Days small projects. Additionally, the conversion of our main APIs to using image_span means that much of the documentation should be updated to be using the new preferred APIs.

**Comment by starsurfer87** [comment]
I would also like to contribute to this as part of Dev Days! I was also looking at the Image Arithmetic section of ImageBufAlgo. Does it  work if I take the chunk starting with `mad` and ending with `channel_sum` so as not to conflict with @filmkorn?

**Comment by lgritz** [comment]
sounds great

**Comment by lgritz** [comment]
Adding a comment to ensure this gets picked up by Clotributor as recently active. This would be a great Dev Days item.

I'll add also that anybody who is converting old examples into tests in this manner should also keep a lookout for if the examples are using raw pointers, and also convert them to use the span and image_span versions of the API calls!

**Comment by marivenkatakrishnan** [comment]
Hi, I’d like to work on this. Is this still open for contribution?

**Comment by lgritz** [comment]
> Hi, I’d like to work on this. Is this still open for contribution?

@marivenkatakrishnan Absolutely. You can choose any particular examples to do the conversion.

Since you are new to the project, I would recommend starting with ONE example. Don't bite off more than you can chew. Once you have it working, pass tests, get approved in code review, then you can take bigger hunks at a time if you want for follow-ups.

**Comment by demiaster** [comment]
Hi @lgritz! If this is still open, I can try to convert one more example in the upcoming DevDay on Thursday. Let me know if it is OK!

**Comment by lgritz** [comment]
@demiaster : Yes, open! Just leave a comment here saying which section of the docs you wish to convert so other people can work on other sections. I think a whole chapter is too much. A section or two of a chapter is good to start, you can always realize partway through the day that it's going really fast and come back and report that you'll do more.

**Comment by demiaster** [comment]
Sounds great, thank you! I'll update this issue tomorrow with details on the section (or better, the part of section) I will tackle :)

**Comment by demiaster** [comment]
I will work on [ImageOutput - Copy an Entire Image](https://openimageio.readthedocs.io/en/master/imageoutput.html#copying-an-entire-image) today, thanks!

**Comment by FaisalXL** [comment]
Hi @lgritz  ! i'm new to PRs I'd like to work on converting the ImageCache chapter example (the ImageCache::create/get_pixels/get_tile snippet in src/doc/imagecache.rst) into a testsuite test! thanks!

**Comment by laurenzhang-dwa** [comment]
Hi @lgritz, I'd like to work on this for the upcoming DevDay on 9/17 (I'll start with [ImageOutput - Channel Names](https://openimageio.readthedocs.io/en/v3.1.16.0/imageoutput.html#channel-names)). I'm not very familiar with C++, so I will probably just be converting the Python snippet. Please let me know if that's alright, thanks!

**Comment by lgritz** [comment]
Yes, that's fine.

**Comment by Yanghaha051218** [comment]
Hi, I'd like to work on one small example from this issue.

I'd like to start with the `ImageBufAlgo::perpixel_op()` C++ example in `src/doc/imagebuf.rst` and convert it into a testsuite example, with the documentation referencing the tested source instead.

Please let me know if this example is still available to work on.

**Comment by lgritz** [comment]
@Yanghaha051218 Yes! Feel free. Note that I have #5449 that converts that part of the docs from rst to md, so beware some shifting files and formats in that area. Hopefully will settle down in the next few days.
